sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs

```diff
@@ -0,0 +1,165 @@
+/*******************************************************************************
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use
+ *  this file except in compliance with the License. A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file.
+ *  This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ *  CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations under the License.
+ * *****************************************************************************
+ *    __  _    _  ___
+ *   (  )( \/\/ )/ __)
+ *   /__\ \    / \__ \
+ *  (_)(_) \/\/  (___/
+ *
+ *  AWS SDK for .NET
+ *  API Version: 2006-03-01
+ *
+ */
+using System;
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.S3.Model;
+
+namespace Amazon.S3.Transfer.Internal
+{
+    /// <summary>
+    /// Buffers downloaded parts in memory using ArrayPool and IPartBufferManager.
+    /// Implements current streaming behavior for multipart downloads.
+    /// </summary>
+    internal class BufferedPartDataHandler : IPartDataHandler
```

Comment:
for downloading to files in the future we can implement the IPartDataHandler logic in a `FilePartDataHandler` and just change the implementation of `ProcessPartAsync`  to write to a file instead of a buffer. This will let use re-use the multi part download coordinator in that implementation